apps/aehttp/src/sc_ws_handler.erl

```diff
@@ -62,35 +96,85 @@ websocket_init(Params) ->
-                    %% At this point, we haven't verified the signature.
-                    %% This is done by the fsm.
-                    lager:debug("Found FSM = ~p", [Fsm]),
-                    case aesc_fsm:reconnect_client(Fsm, self(), SignedTx) of
-                        ok ->
-                            MRef = erlang:monitor(process, Fsm),
-                            { ok, Handler#handler{ fsm_pid  = Fsm
-                                                 , fsm_mref = MRef } };
-                        {error, Err} ->
-                            handler_init_error(Err, Handler)
-                    end
+            case reconnect_to_fsm_(Opts, Handler, fun(E) -> E end) of
+                {ok, _} = Ok ->
+                    Ok;
+                {error, {existing_client, OldClient}} ->
+                    check_existing_client(OldClient, Opts, Handler);
+                {error, Err} ->
+                    handler_init_error(Err, Handler)
             end;
         {error, CheckError} ->
             lager:debug("CheckError = ~p", [CheckError]),
             {stop, undefined}
     end.
 
+reconnect_to_fsm_(#{ channel_id := ChanId
+                   , role       := Role
+                   , pub_key    := Pubkey
+                   , signed_tx  := SignedTx }, Handler, OnError) ->
+    case aesc_fsm:where(ChanId, Role) of
+        undefined ->
+            lager:debug("where(~p, ~p) -> undefined", [ChanId, Role]),
+            maybe_reestablish(Handler, OnError);
+        #{ fsm_pid := Fsm, pub_key := Pubkey } ->
+            %% At this point, we haven't verified the signature.
+            %% This is done by the fsm.
+            lager:debug("Found FSM = ~p", [Fsm]),
+            case aesc_fsm:reconnect_client(Fsm, self(), SignedTx) of
+                ok ->
+                    MRef = erlang:monitor(process, Fsm),
+                    { ok, Handler#handler{ fsm_pid  = Fsm
+                                         , fsm_mref = MRef } };
+                {error, E} ->
+                    OnError(E)
+            end
+    end.
+
+%% The fsm is not running. Perhaps there is cached state, warranting a reestablish?
+%%
+maybe_reestablish(#{ channel_id := ChId
+                   , pub_key    := Pubkey }, OnError) ->
+    case aesc_state_cache:fetch(ChId, Pubkey) of
+        {ok, State, Opts} ->
+            try aesc_offchain_state:get_latest_signed_tx(State) of
+                {_, SignedTx} ->
+                    websocket_init(Opts#{ existing_channel_id => ChId
+                                        , offchain_tx => SignedTx })
+            catch
+                error:_ ->
+                    OnError(not_found)
+            end;
+        error ->
+            OnError(not_found)
+    end.
+
+check_existing_client(Client, Opts, Handler) ->
+    OnError = fun(E) ->
+                      handler_init_error(E, Handler)
+              end,
+    case time_since_last_dispatch(Client) of
+        T when is_integer(T), T < 5000 ->
```

Comment:
The threshold should be defined somewhere at the top and definitely be documented.